aspnetcore/security/authentication/certauth.md

```diff
@@ -206,3 +206,243 @@ Complete the following steps In IIS Manager:
+
+	app.UseEndpoints(endpoints =>
+	{
+		endpoints.MapControllers();
+	});
+}
+```
+
+A separate class can be used to implement validation logic. Because the same self-signed certificate is used in this example, ensure that only your certificate can be used. Validate that the thumbprints of both the client certificate and the server certificate match, otherwise any certificate can be used and will be enough to authenticate. This would be used inside the `AddCertificate` method. You could also validate the subject or the issuer here if you're using intermediate or child certificates.
+
+```csharp
+using System.IO;
+using System.Security.Cryptography.X509Certificates;
+
+namespace AspNetCoreCertificateAuthApi
+{
+    public class MyCertificateValidationService
+    {
+        public bool ValidateCertificate(X509Certificate2 clientCertificate)
+        {
+            var cert = new X509Certificate2(Path.Combine("sts_dev_cert.pfx"), "1234");
+            if (clientCertificate.Thumbprint == cert.Thumbprint)
+            {
+                return true;
+            }
+
+            return false;
+        }
+    }
+}
+```
+
+#### Implement an HttpClient using a certificate
+
+The web API client uses an `HttpClient` which was created using an `IHttpClientFactory` instance. This doesn't provide a way to define a handler for the `HttpClient`, so use an `HttpRequestMessage` to add the certificate to the `X-ARR-ClientCert` request header. The certificate is added as a string using the `GetRawCertDataString` method.
+ 
+```csharp
+private async Task<JsonDocument> GetApiDataAsync()
+{
+	try
+	{
+		var cert = new X509Certificate2(Path.Combine(_environment.ContentRootPath, "sts_dev_cert.pfx"), "1234");
+
+		var client = _clientFactory.CreateClient();
+
+		var request = new HttpRequestMessage()
+		{
+			RequestUri = new Uri("https://localhost:44379/api/values"),
+			Method = HttpMethod.Get,
+		};
+
+		request.Headers.Add("X-ARR-ClientCert", cert.GetRawCertDataString());
+		var response = await client.SendAsync(request);
+
+		if (response.IsSuccessStatusCode)
+		{
+			var responseContent = await response.Content.ReadAsStringAsync();
+			var data = JsonDocument.Parse(responseContent);
+
+			return data;
+		}
+
+		throw new ApplicationException($"Status code: {response.StatusCode}, Error: {response.ReasonPhrase}");
+	}
+	catch (Exception e)
+	{
+		throw new ApplicationException($"Exception {e}");
+	}
+}
+```
+
+If the correct certificate is sent to the server, the data is returned. If no certificate or the wrong certificate is sent, an HTTP 403 status code is returned.
+
+### Create certificates in PowerShell
+
+Creating the certificates is the hardest part in setting up this flow. A self-signed root certificate authority (CA) certificate is created using the `New-SelfSignedCertificate` PowerShell cmdlet. When creating the certificate, use a strong password. It's important to add the `KeyUsageProperty` parameter and the `KeyUsage` parameter as shown.
```

Comment:
You go from self signed to root cas too quickly. Needs an explanation of what the difference is and why you would use one over the other.

Reply:
updated